File: pm/app/src/main/java/cn/partialy/pm/listen/ListenTogetherSocketClient.kt
package cn.partialy.pm.listen

import android.os.SystemClock
import com.google.gson.Gson
import com.google.gson.JsonObject
import io.socket.client.Ack
import io.socket.client.IO
import io.socket.client.Socket
import org.json.JSONObject
import java.util.UUID
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

@Singleton
class ListenTogetherSocketClient @Inject constructor(
    @Named("system_api_base_url") private val systemApiBaseUrl: String,
) {
    interface Listener {
        fun onConnected()
        fun onDisconnected()
        fun onConnectError(message: String)
        fun onLatencyUpdated(latencyMs: Long)
        fun onBroadcast(event: String, raw: JsonObject)
    }

    private val gson = Gson()
    private var socket: Socket? = null
    private var listener: Listener? = null

    val connected: Boolean get() = socket?.connected() == true

    fun connect(token: String, listener: Listener) {
        disconnect()
        this.listener = listener
        val options = IO.Options().apply {
            transports = arrayOf("websocket")
            auth = mapOf("token" to "Bearer $token")
        }
        socket = IO.socket(systemApiBaseUrl, options).apply {
            on(Socket.EVENT_CONNECT) { listener.onConnected() }
            on(Socket.EVENT_DISCONNECT) { listener.onDisconnected() }
            on(Socket.EVENT_CONNECT_ERROR) { args ->
                listener.onConnectError(args.firstOrNull()?.toString().orEmpty().ifBlank { "Socket 连接失败" })
            }
            BROADCAST_EVENTS.forEach { event ->
                on(event) { args ->
                    args.firstOrNull()?.toJsonObject()?.let { listener.onBroadcast(event, it) }
                }
            }
            connect()
        }
    }

    fun disconnect() {
        socket?.apply {
            off()
            disconnect()
            close()
        }
        socket = null
        listener = null
    }

    fun emitJoin(roomId: String, ack: (ListenTogetherAck<ListenTogetherJoinAckData>) -> Unit) {
        emit("listen:join", roomId, "JOIN", emptyMap(), ListenTogetherJoinAckData::class.java, ack)
    }

    fun emitLeave(roomId: String, ack: (ListenTogetherAck<Unit>) -> Unit) {
        emit("listen:leave", roomId, "LEAVE", emptyMap(), Unit::class.java, ack)
    }

    fun emitPlay(
        roomId: String,
        song: ListenTogetherSong,
        position: Long,
        ack: (ListenTogetherAck<ListenTogetherRoomAckData>) -> Unit,
    ) {
        emit("listen:play", roomId, "PLAY", mapOf("song" to song, "position" to position), ListenTogetherRoomAckData::class.java, ack)
    }

    fun emitPause(roomId: String, position: Long, ack: (ListenTogetherAck<ListenTogetherRoomAckData>) -> Unit) {
        emit("listen:pause", roomId, "PAUSE", mapOf("position" to position), ListenTogetherRoomAckData::class.java, ack)
    }

    fun emitChangeSong(
        roomId: String,
        song: ListenTogetherSong,
        autoPlay: Boolean,
        ack: (ListenTogetherAck<ListenTogetherRoomAckData>) -> Unit,
    ) {
        emit("listen:change_song", roomId, "CHANGE_SONG", mapOf("song" to song, "autoPlay" to autoPlay), ListenTogetherRoomAckData::class.java, ack)
    }

    fun emitSeek(roomId: String, position: Long, ack: (ListenTogetherAck<ListenTogetherRoomAckData>) -> Unit) {
        emit("listen:seek", roomId, "SEEK", mapOf("position" to position), ListenTogetherRoomAckData::class.java, ack)
    }

    fun emitEnded(roomId: String, position: Long, ack: (ListenTogetherAck<ListenTogetherRoomAckData>) -> Unit) {
        emit("listen:ended", roomId, "ENDED", mapOf("position" to position), ListenTogetherRoomAckData::class.java, ack)
    }

    fun emitSync(roomId: String, ack: (ListenTogetherAck<ListenTogetherJoinAckData>) -> Unit) {
        emit("listen:sync", roomId, "SYNC", emptyMap(), ListenTogetherJoinAckData::class.java, ack)
    }

    fun emitUpdateRoom(
        roomId: String,
        memberOperation: Boolean,
        ack: (ListenTogetherAck<ListenTogetherRoomAckData>) -> Unit,
    ) {
        emit(
            "listen:update_room",
            roomId,
            "UPDATE_ROOM",
            mapOf("memberOperation" to memberOperation),
            ListenTogetherRoomAckData::class.java,
            ack,
        )
    }

    fun emitQueueEvent(
        roomId: String,
        kind: String,
        targetUserId: String? = null,
        data: Map<String, Any?> = emptyMap(),
        ack: (ListenTogetherAck<Unit>) -> Unit = {},
    ) {
        val eventData = data.toMutableMap().apply {
            put("kind", kind)
            if (!targetUserId.isNullOrBlank()) put("targetUserId", targetUserId)
        }
        emit("listen:queue", roomId, "QUEUE", eventData, Unit::class.java, ack)
    }

    private fun <T> emit(
        event: String,
        roomId: String,
        action: String,
        data: Map<String, Any?>,
        dataClass: Class<T>,
        ack: (ListenTogetherAck<T>) -> Unit,
    ) {
        val socket = socket ?: run {
            ack(ListenTogetherAck(success = false, code = -1, msg = "Socket 未连接", errorMsg = "SOCKET_DISCONNECTED"))
            return
        }
        val payload = ListenTogetherSocketPayload(
            requestId = UUID.randomUUID().toString(),
            roomId = roomId,
            action = action,
            data = data,
        )
        val sentAt = SystemClock.elapsedRealtime()
        socket.emit(event, JSONObject(gson.toJson(payload)), Ack { args ->
            listener?.onLatencyUpdated((SystemClock.elapsedRealtime() - sentAt).coerceAtLeast(0L))
            ack(parseAck(args.firstOrNull(), dataClass))
        })
    }

    private fun <T> parseAck(raw: Any?, dataClass: Class<T>): ListenTogetherAck<T> {
        val json = raw.toJsonObject() ?: return ListenTogetherAck(success = false, code = -1, msg = "服务端响应无效")
        val success = json.get("success")?.asBoolean == true
        val code = json.get("code")?.asInt ?: -1
        val msg = json.get("msg")?.asString.orEmpty()
        val errorMsg = json.get("errorMsg")?.asString
        val data = json.get("data")?.takeIf { it.isJsonObject && dataClass != Unit::class.java }
            ?.let { runCatching { gson.fromJson(it, dataClass) }.getOrNull() }
        return ListenTogetherAck(success = success, code = code, msg = msg, data = data, errorMsg = errorMsg)
    }

    private fun Any?.toJsonObject(): JsonObject? {
        return when (this) {
            is JSONObject -> runCatching { gson.fromJson(toString(), JsonObject::class.java) }.getOrNull()
            is JsonObject -> this
            is String -> runCatching { gson.fromJson(this, JsonObject::class.java) }.getOrNull()
            else -> runCatching { gson.fromJson(gson.toJson(this), JsonObject::class.java) }.getOrNull()
        }
    }

    companion object {
        private val BROADCAST_EVENTS = listOf(
            "ROOM_STATE_CHANGED",
            "ROOM_UPDATED",
            "MEMBER_JOINED",
            "MEMBER_LEFT",
            "MEMBER_KICKED",
            "HOST_TRANSFERRED",
            "ROOM_DESTROYED",
            "QUEUE_EVENT",
            "ERROR_MESSAGE",
        )
    }
}

File: yixi/electron/listenTogether/listenTogetherSocketClient.ts
// 一起听 Socket.IO 客户端（main 进程，单例连接）。
// 鉴权 token 只在 main 内读取；renderer 通过 typed IPC 发送受限命令并取得 ACK + 真实 RTT。

import { randomUUID } from "node:crypto";
import { io, type Socket } from "socket.io-client";
import type {
  ListenTogetherAck,
  ListenTogetherBroadcast,
  ListenTogetherBroadcastEvent,
  ListenTogetherCommandResult,
  ListenTogetherSocketCommand,
} from "../../src/types/listenTogether";
import { getSystemBaseUrl } from "../system/systemClient";

export type ListenTogetherSocketListener = {
  onConnected: () => void;
  onDisconnected: (reason: string) => void;
  onConnectError: (message: string) => void;
  onBroadcast: (event: ListenTogetherBroadcastEvent, message: ListenTogetherBroadcast) => void;
};

const BROADCAST_EVENTS: ListenTogetherBroadcastEvent[] = [
  "ROOM_STATE_CHANGED",
  "ROOM_UPDATED",
  "MEMBER_JOINED",
  "MEMBER_LEFT",
  "MEMBER_KICKED",
  "HOST_TRANSFERRED",
  "ROOM_DESTROYED",
  "QUEUE_EVENT",
  "ERROR_MESSAGE",
];

/** ACK 等待上限：超时转成明确错误，不允许永久 pending */
const ACK_TIMEOUT_MS = 10_000;

let socket: Socket | null = null;

export function connectListenTogetherSocket(token: string, listener: ListenTogetherSocketListener): void {
  // 单例连接：重复 connect 前先清理旧监听和旧 socket
  disconnectListenTogetherSocket();
  const client = io(getSystemBaseUrl(), {
    transports: ["websocket"],
    auth: { token: `Bearer ${token}` },
  });
  client.on("connect", () => listener.onConnected());
  client.on("disconnect", (reason) => listener.onDisconnected(String(reason ?? "")));
  client.on("connect_error", (error) => {
    const message = error instanceof Error ? error.message : String(error ?? "");
    listener.onConnectError(message || "一起听连接失败");
  });
  for (const event of BROADCAST_EVENTS) {
    client.on(event, (message: ListenTogetherBroadcast) => listener.onBroadcast(event, message));
  }
  socket = client;
}

export function disconnectListenTogetherSocket(): void {
  if (!socket) return;
  socket.removeAllListeners();
  socket.disconnect();
  socket = null;
}

export function isListenTogetherSocketConnected(): boolean {
  return socket?.connected === true;
}

type EmitPayload = {
  event: string;
  body: {
    requestId: string;
    roomId: string;
    action: string;
    data: Record<string, unknown>;
  };
};

function failureAck(
  code: number,
  msg: string,
  errorMsg: "SOCKET_DISCONNECTED" | "ACK_TIMEOUT" | "INTERNAL_ERROR",
): ListenTogetherAck {
  return { success: false, code, msg, errorMsg, data: null };
}

/** 把受限命令联合映射为 socket 事件与 PM 同构 payload（{ requestId, roomId, action, data }） */
function buildEmitPayload(command: ListenTogetherSocketCommand): EmitPayload | null {
  const requestId = randomUUID();
  const base = (event: string, action: string, data: Record<string, unknown>): EmitPayload => ({
    event,
    body: { requestId, roomId: command.roomId, action, data },
  });
  switch (command.type) {
    case "join":
      return base("listen:join", "JOIN", {});
    case "leave":
      return base("listen:leave", "LEAVE", {});
    case "sync":
      return base("listen:sync", "SYNC", {});
    case "play":
      return base("listen:play", "PLAY", {
        song: command.song,
        songRef: { source: command.song.source, id: command.song.id },
        position: command.position,
        ...(command.transitionId ? { transitionId: command.transitionId } : {}),
      });
    case "pause":
      return base("listen:pause", "PAUSE", {
        songRef: command.songRef,
        position: command.position,
        ...(command.transitionId ? { transitionId: command.transitionId } : {}),
      });
    case "changeSong":
      return base("listen:change_song", "CHANGE_SONG", {
        song: command.song,
        autoPlay: command.autoPlay,
        ...(command.transitionId ? { transitionId: command.transitionId } : {}),
        ...(command.queueItemId ? { queueItemId: command.queueItemId } : {}),
      });
    case "seek":
      return base("listen:seek", "SEEK", {
        songRef: command.songRef,
        position: command.position,
        ...(command.transitionId ? { transitionId: command.transitionId } : {}),
      });
    case "ended":
      return base("listen:ended", "ENDED", {
        songRef: command.songRef,
        position: command.position,
        ...(command.transitionId ? { transitionId: command.transitionId } : {}),
      });
    case "updateRoom": {
      // 服务端按字段存在性局部更新，未传字段不能出现在 data 中
      const data: Record<string, unknown> = {};
      if (command.roomName !== undefined) data.roomName = command.roomName;
      if (command.maxPeople !== undefined) data.maxPeople = command.maxPeople;
      if (command.memberOperation !== undefined) data.memberOperation = command.memberOperation;
      return base("listen:update_room", "UPDATE_ROOM", data);
    }
    case "queue":
      return base("listen:queue", "QUEUE", {
        ...(command.data ?? {}),
        kind: command.kind,
        ...(command.targetUserId ? { targetUserId: command.targetUserId } : {}),
      });
    case "kickMember":
      return base("listen:kick_member", "KICK_MEMBER", { targetUserId: command.targetUserId });
    case "transferHost":
      return base("listen:transfer_host", "TRANSFER_HOST", { targetUserId: command.targetUserId });
    default:
      return null;
  }
}

function normalizeAck(raw: unknown): ListenTogetherAck {
  if (typeof raw !== "object" || raw === null || Array.isArray(raw)) {
    return failureAck(-1, "服务端响应无效", "INTERNAL_ERROR");
  }
  const value = raw as Record<string, unknown>;
  return {
    success: value.success === true,
    code: typeof value.code === "number" ? value.code : -1,
    msg: typeof value.msg === "string" ? value.msg : "",
    data: value.data,
    ...(typeof value.errorMsg === "string" ? { errorMsg: value.errorMsg } : {}),
  } as ListenTogetherAck;
}

export async function emitListenTogetherCommand(
  command: ListenTogetherSocketCommand,
): Promise<ListenTogetherCommandResult> {
  const client = socket;
  if (!client || !client.connected) {
    return { ack: failureAck(-1, "一起听连接已断开", "SOCKET_DISCONNECTED"), rttMs: null };
  }
  const payload = buildEmitPayload(command);
  if (!payload) {
    return { ack: failureAck(-1, "不支持的一起听命令", "INTERNAL_ERROR"), rttMs: null };
  }
  const sentAt = performance.now();
  try {
    const raw = await client.timeout(ACK_TIMEOUT_MS).emitWithAck(payload.event, payload.body);
    const rttMs = Math.max(0, Math.round(performance.now() - sentAt));
    return { ack: normalizeAck(raw), rttMs };
  } catch {
    return { ack: failureAck(-1, "一起听操作超时", "ACK_TIMEOUT"), rttMs: null };
  }
}
